cmd/reconciler/ip_test.go

```diff
@@ -0,0 +1,247 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/dougbtv/whereabouts/pkg/api/v1alpha1"
+	"github.com/dougbtv/whereabouts/pkg/reconciler"
+	"github.com/dougbtv/whereabouts/pkg/types"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+)
+
+var _ = Describe("Whereabouts IP reconciler", func() {
+	const (
+		ipRange     = "10.10.10.0/16"
+		namespace   = "testns"
+		networkName = "net1"
+		podName     = "pod1"
+	)
+
+	var (
+		reconcileLooper *reconciler.ReconcileLooper
+	)
+
+	BeforeEach(func() {
+		var err error
+		reconcileLooper, err = reconciler.NewReconcileLooper(kubeConfigPath)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	Context("a single running pod", func() {
+		var pod *v1.Pod
+
+		BeforeEach(func() {
+			var err error
+			pod, err = k8sClientSet.CoreV1().Pods(namespace).Create(generatePod(namespace, podName, networkName))
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		Context("with IP from a single IPPool", func() {
+			const poolName = "pool1"
+
+			var pool *v1alpha1.IPPool
+
+			BeforeEach(func() {
+				pool = generateIPPoolSpec(ipRange, namespace, poolName, pod.Name)
+				Expect(k8sClient.Create(context.Background(), pool)).NotTo(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				Expect(k8sClient.Delete(context.Background(), pool)).NotTo(HaveOccurred())
+			})
+
+			Context("the pod remains alive", func() {
+				AfterEach(func() {
+					Expect(k8sClientSet.CoreV1().Pods(namespace).Delete(pod.Name, &metav1.DeleteOptions{})).NotTo(HaveOccurred())
+				})
+
+				It("preserves the pod IP allocation", func() {
+					Expect(reconcileLooper.FindOrphanedIPsPerPool()).To(BeEmpty())
+				})
+			})
+
+			Context("the pod dies", func() {
+				BeforeEach(func() {
+					Expect(k8sClientSet.CoreV1().Pods(namespace).Delete(pod.Name, &metav1.DeleteOptions{})).NotTo(HaveOccurred())
+				})
```

Comment:
Now that I think about it, I probably shouldn't even start up the pod; just the pool pointing at it.